### DISCUSSION TITLE
Modify issue#565 - Do not render questions in HTML documents.

### DISCUSSION THREAD

**Discussion by abing22333** [TongchengOpenSource/smart-doc#567] [Q&A] [answered]
[issue#565](https://github.com/smart-doc-group/smart-doc/issues/565)

### Bug localization
Upon inspecting the /template/html/index.html template, two differences were identified as follows:
查看/template/html/index.html模板，发现两处区别如下：

#### Template code that properly renders HTML formatting.

```javascript
  <div class="paragraph"><p><strong>Description:</strong>${lineBreaksToBr(doc.detail)}</p></div>
```

#### Template code that improperly renders HTML formatting.
```javascript
<td class="tableblock halign-left valign-top"><p class="tableblock">${htmlEscape(param.desc)}</p></td>
```

Continue to trace into the Java code `HtmlEscape.class` and `LineBreaksToBr.class`.
继续跟进到java代码 `HtmlEscape.class`和`LineBreaksToBr.class`

#### The LineBreaksToBr.class that properly renders HTML formatting.
```java
    public String call(Object[] paras, Context ctx) {
        String str = String.valueOf(paras[0]);
        return HtmlUtil.lineBreaksToBr(str);
    }
```

#### The HtmlEscape.class that improperly renders HTML formatting.
```java
    public String call(Object[] paras, Context ctx) {
        String str = String.valueOf(paras[0]).replaceAll("&", "&amp;");
        str = str.replaceAll("\"", "&quot;");
        str = str.replaceAll("<p>", "").replaceAll("</p>", " ");
        return DocUtil.replaceNewLineToHtmlBr(DocUtil.getEscapeAndCleanComment(str));
    }
```

#### Reason for the exception
The `DocUtil.getEscapeAndCleanComment()` called in `HtmlEscape.call()` converts all '>' and '<' characters into their respective HTML entities, preventing any HTML tags from being rendered as elements.

HtmlEscape.call() 中调用的 DocUtil.getEscapeAndCleanComment()。 会将所有的大于小于符合变成转移字符，导致所有的html标签都无法变成html元素.

查看了git提交记录，发现是在commit 0c8bc045 调用了该方法，自此smart-doc就失去了**字段**上html格式渲染的能力。

```java
update: Optimized handling of special greater-than and less-than symbols in documents 

0c8bc045 shalousun <smartdoc647@gmail.com> on 2022/4/3 at 14:46
```

#### Remove the call to DocUtil.getEscapeAndCleanComment().
The simplest solution is to remove the call. However, it's not possible to assess the impact of its removal (I don't understand the purpose of this optimization).
最简单的方案就是移除该调用。不过无法评估移除后的影响(没看懂这个优化的意义)。

#### Refer to the implementation of JavaDoc 
Referencing JavaDoc's support for HTML elements, further reference will be made later, followed by a proposed solution.
参考javadoc对html元素的支持，后面参考了，再出解决方案。

**Comment by shalousun** [comment, marked as answer]
This issue stems from historical reasons; smart-doc has actually been around for several years. In its early days, some methods were developed to address issues with rendering Markdown. For HTML, it's necessary to conduct a series of tests because an improper element might cause problems with the webpage. HTML testing can be aligned with what JavaDoc HTML supports. As long as the page functions correctly, it's fine. It's feasible to individually test and modify Beetl's custom functions for various template issues.
这个是一些历史原因，smart-doc其实好几年了，早期的时候有的是为了解决早期markdown渲染问题编写的方法。html需要写一些测试，某一元素不合理可能会导致页面有问题。html可以按照javadoc html能支持的做些测试。保证页面不出问题就可以，可以单独对各种模版的问题测一测和改改beetl的自定义函数
